apiserver/facades/client/action/action_test.go

```diff
@@ -4,31 +4,552 @@
+		gomock.Any(),
+		coreaction.UUID(actionUUID.String()),
+	).Return(resAction, nil)
+
+	result, err := s.client.Actions(context.Background(), params.Entities{
+		Entities: []params.Entity{
+			{Tag: names.NewActionTag(actionUUID.String()).String()},
+		},
+	})
+
+	c.Assert(err, tc.ErrorIsNil)
+	c.Assert(result.Results, tc.HasLen, 1)
+	c.Assert(result.Results[0].Error, tc.IsNil)
+}
+
+func (s *actionSuite) TestActionsPermissionDenied(c *tc.C) {
+	defer s.setupMocks(c).Finish()
+
+	actionUUID := modeltesting.GenModelUUID(c)
+	// Use a non-admin user tag to fail permission check.
+	nonAdminUser := names.NewUserTag("unauthorized")
+	s.setupAPI(c, nonAdminUser)
+
+	_, err := s.client.Actions(context.Background(), params.Entities{
+		Entities: []params.Entity{
+			{Tag: names.NewActionTag(actionUUID.String()).String()},
+		},
+	})
+
+	c.Assert(err, tc.ErrorIs, apiservererrors.ErrPerm)
+}
+
+func (s *actionSuite) TestActionsInvalidActionTag(c *tc.C) {
+	defer s.setupMocks(c).Finish()
+
+	s.setupAPI(c, jujutesting.AdminUser)
+
+	result, err := s.client.Actions(context.Background(), params.Entities{
+		Entities: []params.Entity{
+			{Tag: "invalid-tag"},
+		},
+	})
+
+	c.Assert(err, tc.ErrorIsNil)
+	c.Assert(result.Results, tc.HasLen, 1)
+	c.Assert(result.Results[0].Error, tc.NotNil)
+	c.Assert(result.Results[0].Error.Code, tc.Equals, params.CodeNotFound)
+}
+
+func (s *actionSuite) TestActionsActionNotFound(c *tc.C) {
+	defer s.setupMocks(c).Finish()
+
+	s.setupAPI(c, jujutesting.AdminUser)
+	actionUUID := modeltesting.GenModelUUID(c)
+
+	s.operationService.EXPECT().GetAction(
+		gomock.Any(),
+		coreaction.UUID(actionUUID.String()),
+	).Return(operation.Action{}, operationerrors.ActionNotFound)
+
+	result, err := s.client.Actions(context.Background(), params.Entities{
+		Entities: []params.Entity{
+			{Tag: names.NewActionTag(actionUUID.String()).String()},
+		},
+	})
+
+	c.Assert(err, tc.ErrorIsNil)
+	c.Assert(result.Results, tc.HasLen, 1)
+	c.Assert(result.Results[0].Error, tc.NotNil)
+	c.Assert(result.Results[0].Error.Code, tc.Equals, params.CodeNotFound)
+}
+
+func (s *actionSuite) TestActionsServerError(c *tc.C) {
+	defer s.setupMocks(c).Finish()
+
+	s.setupAPI(c, jujutesting.AdminUser)
+	actionUUID := modeltesting.GenModelUUID(c)
+
+	s.operationService.EXPECT().GetAction(
+		gomock.Any(),
+		coreaction.UUID(actionUUID.String()),
+	).Return(operation.Action{}, errors.New("boom"))
+
+	result, err := s.client.Actions(context.Background(), params.Entities{
+		Entities: []params.Entity{
+			{Tag: names.NewActionTag(actionUUID.String()).String()},
+		},
+	})
+
+	c.Assert(err, tc.ErrorIsNil)
+	c.Assert(result.Results, tc.HasLen, 1)
+	// This error was already (wrongly) black-holed into a ErrBadId.
+	c.Assert(result.Results[0].Error.Code, tc.Equals, params.CodeNotFound)
+}
+
+func (s *actionSuite) TestActionsMultipleEntities(c *tc.C) {
+	defer s.setupMocks(c).Finish()
+
+	s.setupAPI(c, jujutesting.AdminUser)
+	actionUUID0 := modeltesting.GenModelUUID(c)
```

Comment:
todo: use `uuid.MustNewUUID()` through out the file, we're not using ModelUUIDs in operation code.

Reply:
Done.